synapse/app/_base.py

```diff
@@ -467,12 +467,14 @@ def run_sighup(*args: Any, **kwargs: Any) -> None:
     # everything currently allocated are things that will be used for the
     # rest of time. Doing so means less work each GC (hopefully).
     #
-    gc.collect()
-    gc.freeze()
-
-    # Speed up shutdowns by freezing all allocated objects. This moves everything
-    # into the permanent generation and excludes them from the final GC.
-    atexit.register(gc.freeze)
+    # gc.freeze may not be available in all Python implementations, thus we check here
```

Comment:
It may be useful to actually include the string "pypy" in here somewhere, since that's the most common alternative to CPython that we care about... sort of a breadcrumb for us to find our way back with later.

```suggestion
    # PyPy does not (yet?) implement gc.freeze()
```